src/templates/main.wxs.mustache

```diff
@@ -21,12 +21,16 @@
   destination for 32-bit versus 64-bit installers. Removal of these lines will
   cause installation errors.
 -->
-<?if $(var.Platform) = x64 ?>
-    <?define Win64 = "yes" ?>
+<?if $(sys.BUILDARCH) = x64 or $(sys.BUILDARCH) = arm64 ?>
     <?define PlatformProgramFilesFolder = "ProgramFiles64Folder" ?>
 <?else ?>
-  <?define Win64 = "no" ?>
-  <?define PlatformProgramFilesFolder = "ProgramFilesFolder" ?>
+    <?define PlatformProgramFilesFolder = "ProgramFilesFolder" ?>
+<?endif ?>
+
+<?ifdef CargoTargetExplicit ?>
+    <?define CargoTargetPathName = "$(var.CargoTarget)" ?>
+<?else ?>
+    <?define CargoTargetPathName = "" ?>

```

Comment:
I do wonder, can we just collapse the `CargoTarget`, `CargoTargetPathName`, and `Profile` variables down into a single variable? Then, in the cargo-wix _create_ implementation, we build the "full" path to the binary? I think this would avoid needing the `CargoTargetExplicit` hack. We don't allow overriding the `Cargo` variables in the CLI...yet. 

Partly, I am just really confused by all of this, which is probably due to my lack of understanding about the cross compilation process with Cargo. So, please humor me with some external thoughts and ramblings:

At the moment, I have the following WiX variables that are all related to the binary source path in the WXS file:

1. `CargoTargetDir`
2. `CargoTarget`
3. `CargoTargetExplicit`
4. `CargoTargetPathName`
5. `Profile` (Probably should be renamed to `CargoProfile` to be consistent)

These get joined together in the printing/initialization as:

`$(var.CargoTargetDir)\$(var.CargoTargetPathName)\$(var.Profile)\Example.exe`

and `CargoTarget` does not exist in the path. However, in the preprocessor, the `CargoTarget` is used for the `CargoTargetPathName` if the `CargoTargetExplicit` variable is defined; otherwise, an empty path is used. Confusingly, a WiX variable is used as the value for the `CargoTargetPathName`. Can this preprocessor logical be moved to the Rust implementation and have a single `$(var.TargetBinaryDirectory)` or similar name variable? Then,

```rust
.arg(format!("-dTargetBinaryDirectory={}", target_binary_directory));
```

where `target_binary_directory` is determined with a method earlier in the `run` method. Maybe this conflicts with the _init_ and _print_ subcommands? 

The printed WXS template would be:

```xml
Source="$(var.TargetBinaryDirectory)\Example.exe"
```

where the `Example.exe` is set during the init/print. A CLI override could be added for the `TargetBinaryDirectory` for greater flexibility. I just don't know if the individual path components need to be separate WiX variables. I don't think you are able to change these path components when building a project in Cargo. Is this assumption correct?

A little background because I could not picture the cross compilation process: for each target, i686 and x86_64, Cargo creates a separate target directory, but only if the `--target` option is used; otherwise, it defaults to a profile folder. So, the binary source paths can be:

1. `target\debug\Example.exe`
2. `target\release\Example.exe`
3. `target\i686-pc-windows-msvc\debug\Example.exe`
4. `target\i686-pc-windows-msvc\release\Example.exe`
5. `target\x86_64-pc-windows-msvc\debug\Example.exe`
6. `target\x86_64-pc-windows-msvc\release\Example.exe`

assuming the target directory from the `cargo metadata` is `target`.

Reply:
> A little background because I could not picture the cross compilation process: for each target, i686 and x86_64, Cargo creates a separate target directory, but only if the --target option is used; otherwise, it defaults to a profile folder. So, the binary source paths can be:

Yes, this is correct. I probably could move the whole logic into Rust code. The reasoning behind the current system is that I want to pass CargoTarget to the WXS so that it can be used independently. For instance, CargoTarget could be used to discriminate between -gnu and -msvc toolchains from the WXS, allowing the user to change the title or chose a different set of DLL dependencies to install.

So with CargoTarget being passed, I figured I could generate the full target path in a dynamic way. But you're probably right that instead of making the WXS generate the full path to the generated binary, it'd make more sense to let the Rust side generate it. It'd be more future-proof too, if we need to make further changes to resolve where exes are located without breaking user's code.

Here's what I could do: I'll keep CargoTarget, CargoTargetDir and Profile, as those are useful primitives for the user to use to customize their installers. I'll remove CargoTargetExplicit, and add a new CargoTargetBinDir that the rust code will generate to point to `CargoTargetDir\(CargoTarget\)?Profile\`. This way, the full set of variables sent to the WXS will be:

- Profile: The name of the Cargo profile used to build the binary. Either debug or release.
- Version: Version of the program we're installing.
- Platform (deprecated): An alias for sys.BUILDARCH, One of: x86, x64, arm, arm64
- CargoTargetDir: The path to the Cargo Target directory, equivalent to `target-dir` cargo config.
- CargoTarget: The name of the cargo target used to build the binary.
- CargoTargetBinDir: The full path where cargo will store the final binaries.

Those all seem to be independently useful. What do you think? (Also, I think having ^ documented as comments in the WXS template could be really useful for people writing the WXS ^^).

Reply:
>Yes, this is correct. I probably could move the whole logic into Rust code. The reasoning behind the current system is that I want to pass CargoTarget to the WXS so that it can be used independently. For instance, CargoTarget could be used to discriminate between -gnu and -msvc toolchains from the WXS, allowing the user to change the title or chose a different set of DLL dependencies to install.

Ah, I am now getting it! The embedded WXS template may not use these variables, but a user may want access to this information in his or her own customized WXS file. Basically, cargo-wix should be providing Cargo and project build-related variables as WiX variables, like the [Cargo environment variables](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables) provided to [crates](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-crates) and [build scripts](https://doc.rust-lang.org/cargo/reference/environment-variables.html#environment-variables-cargo-sets-for-build-scripts). Got it! This is totally cool and inline with both the cargo-wix subcommand and the Cargo build process. Love it! 

This is actually a separate feature than implementing cross-compilation. I have created #118 to track and discuss a more complete implementation. We can use this PR with its initial variables as a foundation to expand out the feature. 

>Those all seem to be independently useful. What do you think? If that works for you, I'll update this PR asap.

Yes. Please implement moving the logic into Rust and the mentioned variables. I think the `Profile` variable will eventually need to be renamed to `CargoProfile` for consistency and depending on the naming convention. However, this might be a change similar to removing the `Platform` variable. I have created #121 to track deprecation and eventual removal of the `Platform` variable, and similarly I have created #122 to track renaming the `Profile` variable.

>(Also, I think having ^ documented as comments in the WXS template could be really useful for people writing the WXS).

Yes! Some comments at the top of the WXS template about the WiX variables that are provided automatically by cargo-wix should be added. A `WiX Variables` section/heading should also be added to the online help documentation, too. I have created #119 and #120 to track these separately from this PR.

I wonder if some kind of "show variables" subcommand/argument/flag/option would be useful that also lists all of the variables provided? Basically, some way to add short description of each cargo-wix-provided WiX variable like the `-h,--help` content for the CLI?

Reply:
I moved CargoTargetBinDir resolution to the rust side.